File: web/src/domain/__tests__/ufEngine.test.ts
import { describe, expect, it } from "vitest";
import { ZERO, money } from "../money.ts";
import { of } from "../dates.ts";
import {
  annualisedPct, clpToUf, currentOf, delta, deltaPct, downsample, futureOf, historyWindow, ufToClp,
} from "../ufEngine.ts";
import { expectWithin, num } from "../testing.ts";

const today = of(2026, 9, 5);
const series = [
  { date: of(2026, 9, 3), value: money("40877.73") },
  { date: of(2026, 9, 4), value: money("40879.04") },
  { date: of(2026, 9, 5), value: money("40880.36") },
  { date: of(2026, 9, 6), value: money("40881.68") },
  { date: of(2026, 9, 9), value: money("40885.63") },
];

describe("uf engine", () => {
  it("current value ignores published future dates", () => {
    const current = currentOf(series, today);

    expect(current).not.toBeNull();
    expect(current!.date).toBe(today);
    expect(current!.value.toString()).toBe("40880.36");
  });

  it("future values are the ones beyond today", () => {
    const future = futureOf(series, today);

    expect(future).toHaveLength(2);
    expect(future[0]!.date).toBe(of(2026, 9, 6));
  });

  it("conversions round trip", () => {
    const rate = money("40880.36");
    const uf = clpToUf(money("1000000"), rate);

    expectWithin(num(uf), 0.0001, 1000000 / 40880.36);
    expectWithin(num(ufToClp(uf, rate)), 5, 1000000);
  });

  it("conversion with a zero rate does not divide by zero", () => {
    expect(clpToUf(money("1000"), ZERO).toString()).toBe("0");
    expect(deltaPct(ZERO, money("10")).toString()).toBe("0");
  });

  it("deltas carry the right sign", () => {
    expect(delta(money("100"), money("102")).toString()).toBe("2");
    expectWithin(num(deltaPct(money("100"), money("102"))), 1e-9, 2);
    expectWithin(num(deltaPct(money("100"), money("98"))), 1e-9, -2);
  });

  it("an empty series has no current value", () => {
    expect(currentOf([], today)).toBeNull();
    expect(futureOf([], today)).toHaveLength(0);
  });

  // ------------------------------------------------------------ annualisation

  it("a full year annualises to its own change", () => {
    expectWithin(num(annualisedPct(money("100"), money("104"), 365)), 0.05, 4);
  });

  it("a half year annualises to roughly the compounded rate", () => {
    expectWithin(num(annualisedPct(money("100"), money("102"), 183)), 0.1, 4.04);
  });

  it("a decline annualises negative", () => {
    expectWithin(num(annualisedPct(money("100"), money("98"), 365)), 0.05, -2);
  });

  it("annualising a short window magnifies the move, as it should", () => {
    expect(num(annualisedPct(money("100"), money("101"), 30))).toBeGreaterThan(12);
  });

  it("annualisation refuses impossible inputs instead of returning nonsense", () => {
    expect(annualisedPct(money("100"), money("110"), 0).toString()).toBe("0");
    expect(annualisedPct(ZERO, money("110"), 365).toString()).toBe("0");
    expect(annualisedPct(money("100"), ZERO, 365).toString()).toBe("0");
  });

  it("real UF movement annualises close to Chilean inflation", () => {
    const annual = num(annualisedPct(money("39434.60"), money("40880.36"), 365));

    expect(annual).toBeGreaterThan(2);
    expect(annual).toBeLessThan(6);
  });

  // ------------------------------------------------------------- downsampling

  const long = Array.from({ length: 18_000 }, (_, k) => ({
    date: of(1977, 8, 1),
    value: money(String(1000 + k)),
  }));

  it("a short series is returned untouched", () => {
    const short = series.slice(0, 3);
    expect(downsample(short, 400)).toBe(short);
  });

  it("downsampling keeps the endpoints and the requested size", () => {
    const thinned = downsample(long, 400);

    expect(thinned).toHaveLength(400);
    expect(thinned[0]!.value.toString()).toBe(long[0]!.value.toString());
    expect(thinned[399]!.value.toString()).toBe(long[17_999]!.value.toString());
  });

  it("an absurd point budget does not crash", () => {
    expect(downsample(series, 1)).toBe(series);
    expect(downsample([], 400)).toHaveLength(0);
  });

  // ------------------------------------------------------------ history window

  const window = Array.from({ length: 400 }, (_, k) => ({
    date: of(2025, 9, 5),
    value: money(String(40000 + k)),
  })).map((v, k) => ({ ...v, date: shift(of(2025, 9, 5), k) }));

  it("the window never reaches past today", () => {
    const w = historyWindow(window, 3, of(2026, 9, 5));

    expect(w[w.length - 1]!.date).toBe(of(2026, 9, 5));
    expect(w.every((v) => v.date <= of(2026, 9, 5))).toBe(true);
  });

  it("the whole history also stops at today", () => {
    const w = historyWindow(window, null, of(2026, 9, 5));

    expect(w[0]!.date).toBe(of(2025, 9, 5));
    expect(w[w.length - 1]!.date).toBe(of(2026, 9, 5));
  });

  it("the window starts the requested number of months back", () => {
    expect(historyWindow(window, 3, of(2026, 9, 5))[0]!.date).toBe(of(2026, 6, 5));
  });

  it("a series entirely in the future yields nothing to chart", () => {
    const future = [
      { date: of(2026, 9, 6), value: money("40881.68") },
      { date: of(2026, 9, 9), value: money("40885.63") },
    ];
    expect(historyWindow(future, null, of(2026, 9, 5))).toHaveLength(0);
  });

  it("an empty series stays empty", () => {
    expect(historyWindow([], 3)).toHaveLength(0);
  });
});

function shift(start: string, days: number): string {
  const d = new Date(`${start}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

File: app/src/test/java/cl/tickers/app/domain/UfEngineTest.kt
package cl.tickers.app.domain

import cl.tickers.app.domain.engine.UfEngine
import cl.tickers.app.domain.model.DatedValue
import com.google.common.truth.Truth.assertThat
import org.junit.Test
import java.math.BigDecimal
import java.time.LocalDate

class UfEngineTest {

    private val today = LocalDate.of(2026, 9, 5)

    private val series = listOf(
        DatedValue(LocalDate.of(2026, 9, 3), BigDecimal("40877.73")),
        DatedValue(LocalDate.of(2026, 9, 4), BigDecimal("40879.04")),
        DatedValue(LocalDate.of(2026, 9, 5), BigDecimal("40880.36")),
        DatedValue(LocalDate.of(2026, 9, 6), BigDecimal("40881.68")),
        DatedValue(LocalDate.of(2026, 9, 9), BigDecimal("40885.63")),
    )

    @Test
    fun `current value ignores published future dates`() {
        val current = UfEngine.currentOf(series, today)

        assertThat(current).isNotNull()
        assertThat(current!!.date).isEqualTo(today)
        assertThat(current.value).isEqualTo(BigDecimal("40880.36"))
    }

    @Test
    fun `future values are the ones beyond today`() {
        val future = UfEngine.futureOf(series, today)

        assertThat(future).hasSize(2)
        assertThat(future.first().date).isEqualTo(LocalDate.of(2026, 9, 6))
        assertThat(future.map { it.date }).isInOrder()
    }

    @Test
    fun `isFuture marks published-but-not-yet-reached days`() {
        assertThat(series.first().isFuture(today)).isFalse()
        assertThat(series.last().isFuture(today)).isTrue()
    }

    @Test
    fun `conversions round trip`() {
        val rate = BigDecimal("40880.36")
        val uf = UfEngine.clpToUf(BigDecimal("1000000"), rate)
        val back = UfEngine.ufToClp(uf, rate)

        assertThat(uf.toDouble()).isWithin(0.0001).of(1000000 / 40880.36)
        assertThat(back.toDouble()).isWithin(5.0).of(1000000.0)
    }

    @Test
    fun `conversion with a zero rate does not divide by zero`() {
        assertThat(UfEngine.clpToUf(BigDecimal("1000"), BigDecimal.ZERO))
            .isEqualTo(BigDecimal.ZERO)
        assertThat(UfEngine.deltaPct(BigDecimal.ZERO, BigDecimal("10")))
            .isEqualTo(BigDecimal.ZERO)
    }

    @Test
    fun `deltas carry the right sign`() {
        assertThat(UfEngine.delta(BigDecimal("100"), BigDecimal("102")))
            .isEqualTo(BigDecimal("2"))
        assertThat(UfEngine.deltaPct(BigDecimal("100"), BigDecimal("102")).toDouble())
            .isWithin(1e-9).of(2.0)
        assertThat(UfEngine.deltaPct(BigDecimal("100"), BigDecimal("98")).toDouble())
            .isWithin(1e-9).of(-2.0)
    }

    @Test
    fun `an empty series has no current value`() {
        assertThat(UfEngine.currentOf(emptyList(), today)).isNull()
        assertThat(UfEngine.futureOf(emptyList(), today)).isEmpty()
    }

    // --------------------------------------------------------- annualisation

    @Test
    fun `a full year annualises to its own change`() {
        val annual = UfEngine.annualisedPct(BigDecimal("100"), BigDecimal("104"), 365)

        assertThat(annual.toDouble()).isWithin(0.05).of(4.0)
    }

    @Test
    fun `a half year annualises to roughly the compounded rate`() {
        // 2% in six months compounds to about 4.04% a year.
        val annual = UfEngine.annualisedPct(BigDecimal("100"), BigDecimal("102"), 183)

        assertThat(annual.toDouble()).isWithin(0.1).of(4.04)
    }

    @Test
    fun `a decline annualises negative`() {
        val annual = UfEngine.annualisedPct(BigDecimal("100"), BigDecimal("98"), 365)

        assertThat(annual.toDouble()).isWithin(0.05).of(-2.0)
    }

    @Test
    fun `annualising a short window magnifies the move, as it should`() {
        // 1% in a month is a much larger annual rate; the UI labels it so.
        val annual = UfEngine.annualisedPct(BigDecimal("100"), BigDecimal("101"), 30)

        assertThat(annual.toDouble()).isGreaterThan(12.0)
    }

    @Test
    fun `annualisation refuses impossible inputs instead of returning nonsense`() {
        assertThat(UfEngine.annualisedPct(BigDecimal("100"), BigDecimal("110"), 0))
            .isEqualTo(BigDecimal.ZERO)
        assertThat(UfEngine.annualisedPct(BigDecimal.ZERO, BigDecimal("110"), 365))
            .isEqualTo(BigDecimal.ZERO)
        assertThat(UfEngine.annualisedPct(BigDecimal("100"), BigDecimal.ZERO, 365))
            .isEqualTo(BigDecimal.ZERO)
    }

    @Test
    fun `real UF movement annualises close to Chilean inflation`() {
        // 5-sep-2025 to 5-sep-2026 in the real series.
        val annual = UfEngine.annualisedPct(BigDecimal("39434.60"), BigDecimal("40880.36"), 365)

        assertThat(annual.toDouble()).isGreaterThan(2.0)
        assertThat(annual.toDouble()).isLessThan(6.0)
    }

    // -------------------------------------------------------- downsampling

    @Test
    fun `a short series is returned untouched`() {
        val short = series.take(3)

        assertThat(UfEngine.downsample(short, 400)).isSameInstanceAs(short)
    }

    @Test
    fun `downsampling keeps the endpoints and the requested size`() {
        val long = (0 until 18_000).map {
            DatedValue(LocalDate.of(1977, 8, 1).plusDays(it.toLong()), BigDecimal(1000 + it))
        }

        val thinned = UfEngine.downsample(long, 400)

        assertThat(thinned).hasSize(400)
        assertThat(thinned.first()).isEqualTo(long.first())
        assertThat(thinned.last()).isEqualTo(long.last())
        assertThat(thinned.map { it.date }).isInOrder()
    }

    @Test
    fun `downsampling preserves the overall movement`() {
        val long = (0 until 18_000).map {
            DatedValue(LocalDate.of(1977, 8, 1).plusDays(it.toLong()), BigDecimal(1000 + it))
        }
        val thinned = UfEngine.downsample(long, 400)

        assertThat(UfEngine.deltaPct(thinned.first().value, thinned.last().value))
            .isEqualTo(UfEngine.deltaPct(long.first().value, long.last().value))
    }

    @Test
    fun `an absurd point budget does not crash`() {
        assertThat(UfEngine.downsample(series, 1)).isSameInstanceAs(series)
        assertThat(UfEngine.downsample(emptyList(), 400)).isEmpty()
    }

    // ------------------------------------------------------- history window

    private val longSeries = (0L until 400L).map {
        DatedValue(LocalDate.of(2025, 9, 5).plusDays(it), BigDecimal(40000 + it))
    }

    /**
     * The regression this pins: the chart ended on the last published day,
     * which is up to a month in the future, so its final label and its period
     * variation reached past today.
     */
    @Test
    fun `the window never reaches past today`() {
        val today = LocalDate.of(2026, 9, 5)

        val window = UfEngine.historyWindow(longSeries, months = 3, today = today)

        assertThat(window.last().date).isEqualTo(today)
        assertThat(window.none { it.date.isAfter(today) }).isTrue()
    }

    @Test
    fun `the whole history also stops at today`() {
        val today = LocalDate.of(2026, 9, 5)

        val window = UfEngine.historyWindow(longSeries, months = null, today = today)

        assertThat(window.first().date).isEqualTo(LocalDate.of(2025, 9, 5))
        assertThat(window.last().date).isEqualTo(today)
    }

    @Test
    fun `the window starts the requested number of months back`() {
        val today = LocalDate.of(2026, 9, 5)

        val window = UfEngine.historyWindow(longSeries, months = 3, today = today)

        assertThat(window.first().date).isEqualTo(LocalDate.of(2026, 6, 5))
    }

    @Test
    fun `today itself is included`() {
        val today = LocalDate.of(2026, 9, 5)

        assertThat(UfEngine.historyWindow(longSeries, months = 1, today = today).map { it.date })
            .contains(today)
    }

    @Test
    fun `a series entirely in the future yields nothing to chart`() {
        val future = listOf(
            DatedValue(LocalDate.of(2026, 9, 6), BigDecimal("40881.68")),
            DatedValue(LocalDate.of(2026, 9, 9), BigDecimal("40885.63")),
        )

        assertThat(UfEngine.historyWindow(future, months = null, today = LocalDate.of(2026, 9, 5)))
            .isEmpty()
    }

    @Test
    fun `an empty series stays empty`() {
        assertThat(UfEngine.historyWindow(emptyList(), months = 3)).isEmpty()
    }
}
